### PR TITLE
TOOLS-4263 Convert mongorestore oplog-replay edge-case tests to Go

### PR DESCRIPTION
**Pull request by autarch**
Adds five tests in `mongorestore/oplog_edges_test.go` covering mongorestore's oplog-replay edge cases, and moves the BSON fixtures these tests need from `test/qa-tests/jstests/restore/testdata/` to `mongorestore/testdata/`. This also adds `testutil.SkipUnlessStandalone` in `common/testutil/testutil.go`: restoring a dumped `local/oplog.rs.bson` as a collection only works on a standalone, because a replica set rejects direct writes to `local.oplog.rs` and `mongos` rejects writes to the `local` database and does not support `applyOps`. Evergreen's `integration-*-cluster` variants are replica sets, so the affected test skips there.

JS -> Go mapping (all Go tests live in `mongorestore/oplog_edges_test.go`):

- `test/qa-tests/jstests/restore/oplog_replay_conflict.js` -> `TestOplogReplayConflict` - a dump that contains an embedded `oplog.bson` and is also given an explicit `--oplogFile` must fail, and must apply no data at all.
- `test/qa-tests/jstests/restore/oplog_replay_priority_oplog.js` -> `TestOplogReplayPriorityOplog` - when both an embedded oplog and a higher-priority `--oplogFile` are available, only the entries from the `--oplogFile` are applied. Standalone only, per the skip described above.
- `test/qa-tests/jstests/restore/oplog_replay_noop.js` -> `TestOplogReplayNoop` - noop oplog entries are skipped during replay while the surrounding inserts are applied.
- `test/qa-tests/jstests/restore/preserve_oplog_structure_order.js` -> `TestOplogReplayPreservesComplexIDOrder` - an update op whose `_id` is a multi-field subdocument preserves that subdocument's field order through replay, so the op matches the intended document. The document is pre-inserted on server 8.1+ per SERVER-88158.
- `test/qa-tests/jstests/restore/oplog_replay_size_safety.js` -> `TestOplogReplaySizeSafety` - replay batches ops without exceeding the server's message size limit. The matrix is reduced to 50k small ops plus 8 roughly 1MB ops; the JS version swept up to a million ops, which is far slower for the same signal.

### CODE REVIEW

**Comment by autarch** [comment]
* **#1104** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1104?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1102** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1102?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1098** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1098?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1097** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1097?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1096** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1096?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1093** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1093?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1091** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1091?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1074** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1074?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1073** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1073?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1049** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1049?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1060** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1060?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1060?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1046** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1046?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1048](https://github.com/mongodb/mongo-tools/pull/1048) <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1048?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1041** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1041?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1047** <a href="https://app.graphite.com/github/pr/mongodb/mongo-tools/1047?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: mongorestore/oplog_edges_test.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+		OplogReplayOption,
+		OplogFileOption, "testdata/extra_oplog.bson",
+		DirectoryOption, "testdata/dump_oplog_conflict",
+	)
+	require.ErrorContains(
+		t, result.Err, "cannot provide both an oplog.bson file and an oplog file",
+		"providing two top-priority oplogs errors",
+	)
+
+	count, err := coll.CountDocuments(context.Background(), bson.M{})
```
Comment (mmcclimon) [inline mongorestore/oplog_edges_test.go]: (nit agin) Usually we use bson.D here.

File: mongorestore/oplog_edges_test.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	require.NoError(t, err, "counting documents")
+	assert.EqualValues(t, 0, count, "no entries are restored when the oplogs conflict")
+}
+
+// TestOplogReplayPriorityOplog converts oplog_replay_priority_oplog.js: when a
```
Comment (mmcclimon) [inline mongorestore/oplog_edges_test.go]: I don't think these comments that say "this converts whatever.js" are useful.

File: mongorestore/oplog_edges_test.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+// dump directory contains a local/oplog.rs.bson AND --oplogFile points at a
+// higher-priority oplog, only the priority oplog's entries are applied.
+func TestOplogReplayPriorityOplog(t *testing.T) {
+	testtype.SkipUnlessTestType(t, testtype.IntegrationTestType)
+
+	// This test restores the dump_local_oplog fixture, which contains a
+	// local/oplog.rs.bson. Writing local.oplog.rs is only allowed on a
+	// standalone. Like the JS test this replaces, it is only meaningful there.
```
Comment (mmcclimon) [inline mongorestore/oplog_edges_test.go]: Needless reference to JS test.

File: mongorestore/oplog_edges_test.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	aCount, err := coll.CountDocuments(context.Background(), bson.M{"a": 1})
+	require.NoError(t, err, "counting {a:1} documents")
+	assert.EqualValues(t, 1, aCount, "the inserted document has the expected contents")
+}
+
+// TestOplogReplayPreservesComplexIDOrder converts
+// preserve_oplog_structure_order.js: replaying an update op whose _id is a
+// multi-field subdocument must preserve the field order, otherwise the server
+// rejects the op (the o._id and o2._id must match).
+func TestOplogReplayPreservesComplexIDOrder(t *testing.T) {
```
Comment (mmcclimon) [inline mongorestore/oplog_edges_test.go:150]: The JS version of this tests runs restore 10 times; I'm not totally sure why, but the comment suggests that it might randomly do the correct thing, and that running it multiple times will ensure it's _actually_ doing the right thing. Should we keep that behavior here?

**Review by mmcclimon — APPROVED** [review]
Sure, LGTM.

**Comment by autarch** [comment]
### Merge activity

* **Aug 12, 6:36 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/mongodb/mongo-tools/1060).
* **Aug 12, 6:40 PM UTC**: [Graphite](https://app.graphite.com/github/pr/mongodb/mongo-tools/1060) rebased this pull request as part of a merge.
* **Aug 12, 6:40 PM UTC**: @autarch merged this pull request with [Graphite](https://app.graphite.com/github/pr/mongodb/mongo-tools/1060).
